### ISSUE TITLE
Consider operators for controlling "truthiness" of a value.

### ISSUE THREAD

**Issue by rbuckton** [littledan/proposal-operator-overloading#16]
In the explainer it states:

> The following operators do not support overloading:
> - `!`, `&&`, `||` (boolean operations--always does ToBoolean first, and then works with the boolean)

However, it would be valuable to control *how* an object should be converted to a boolean value.

In C#, you can control boolean coercion using [the `true()` and `false()` operators](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/true-false-operators):

```cs
// in C#
struct Foo {
  public value: int;
  public static operator true(Foo x) {
    return x.value === 0;
  }
}
```

**Comment by littledan** [comment]
I'd like to preserve the invariant that objects are always truthy. For value types (a separate, orthogonal proposal), we might consider having falsiness, though I don't know how exactly it should work. Actually calling a method for conversion to a boolean sounds much more dynamic than I was expecting; I believe this would cause issues both for invariants JS programmers may want to count on as well as for implementations. IIRC during the development of BigInt, we considered making ToBoolean throw, but this was deemed to be bad for both of those reasons.

**Comment by rbuckton** [comment]
From @littledan in https://github.com/littledan/proposal-operator-overloading/issues/15#issuecomment-562405856:
> There's a lot here, but just to respond to part of it: As I mentioned in the presentation, it's an explicit non-goal to make changes to primitive type coercion. We already have toString, valueOf, and @@toPrimitive. I could imagine the argument for value types having a non-wrapper/object-based coercion facility (though I'm a bit skeptical that it pays for itself), but this proposal does not add value types, and provides overloading for objects.

My concern is that user-code might not be able to express the same types of coercion performed by built-ins like `Symbol` or `BigInt`. A `Symbol` has a `toString` but `Symbol() + ""` throws.

Also, the `ToBoolean` coercion does not allow for a user-defined implementation to determine truthiness or falsiness, which would be useful for a number-like user-defined type.

**Comment by rbuckton** [comment]
I'd be fine with considering the overloading of primitive coercion a feature of operator overloading tied to a "value types" proposal, but I do believe it is a cross-cutting concern.

**Comment by littledan** [comment]
Yeah, I haven't articulated how this would work in value types, so this would be important to document well as an exercise; at the same time, as far as moving a value types proposal forward, I'd like to start with records and tuples, which wouldn't have this capability by themselves.

**Comment by kmiller68** [comment]
I think @ljharb has library concerns with an `operator bool` overload (using the C++ name for it). It sounded like many libraries use if truthiness as part of their implementation to switch based on type.

**Comment by ljharb** [comment]
Indeed, changing the basic axiom that all objects are truthy would break a lot of generic inspection code and older shim code.

**Comment by hax** [comment]
It seems allowing `bool` overload for objects may have compatibility issue. But it should be ok for value types. And I'm wondering maybe we should only allow operator overloading in value types? It seems current use cases are all like value types and it also avoid the many issues in QA (like subclass issue).

**Comment by ljharb** [comment]
It would only be OK if the value type was a primitive, not an object.

**Comment by hax** [comment]
@ljharb Yeah what I mean is user defined values types which should be very like primitives (`typeof` should not return `object` for them).

**Comment by jhmaster2000** [comment]
> Indeed, changing the basic axiom that all objects are truthy would break a lot of generic inspection code and older shim code.

> I think @ljharb has library concerns with an `operator bool` overload (using the C++ name for it). It sounded like many libraries use if truthiness as part of their implementation to switch based on type.


I don't see how exactly is this breaking any existing code at all given there are zero changes whatsoever to the truthy/falsy behavior without the addition of the `with operators from` statement to the file? All these old libraries and even modern code made after the implementation of this proposal will be completely oblivious of any truthy/falsy discrepancy for any object they have unless they ask for it.

Additionally I would like to propose we simply allow overloading of the `!` operator, enforce it to only return a boolean (Throwing a TypeError otherwise), and then it can be used to infer the values of both truthy and falsy for the object:
```js
class Vector {
  constructor(x = 0, y = 0, z = 0) {
    this.x = x; this.y = y; this.z = z;
  }
  operator!() {
    return (this.x === 0 && this.y === 0 && this.z === 0);
  }
}

if (new Vector(0, 0, 0));  // !Vector[operator!]() => !true => false
if (!new Vector(0, 0, 0)); // Vector[operator!]()  => true
if (new Vector(1, 1, 1));  // !Vector[operator!]() => !false => true
if (!new Vector(1, 1, 1)); // Vector[operator!]()  => false
```
